File: kernel/rust_null.c
// SPDX-License-Identifier: GPL-2.0
/*
 * rust_null — misc device with the Linux /dev/null I/O contract.
 *
 * /dev/null itself is mem 1:3 (drivers/char/mem.c, null_fops). This
 * module registers a *misc* device named "rust-null" (major 10, dynamic
 * minor) whose file_operations copy that contract:
 *
 *   read_null      → return 0
 *   write_null     → return count (do not touch the user buffer)
 *   null_lseek     → file->f_pos = 0; return 0
 *   read_iter_null / write_iter_null
 *
 * The VFS hook is struct file_operations, hung off struct miscdevice
 * via misc_register(). That is the same path Rust-for-Linux
 * kernel::miscdevice::MiscDevice uses (see rfl/rust_null.rs). Distro
 * kernels without CONFIG_RUST cannot compile the .rs; this .c is the
 * loadable module.
 */

#include <linux/fs.h>
#include <linux/miscdevice.h>
#include <linux/module.h>
#include <linux/uio.h>

static ssize_t rust_null_read(struct file *file, char __user *buf,
			      size_t count, loff_t *ppos)
{
	return 0;
}

static ssize_t rust_null_write(struct file *file, const char __user *buf,
			       size_t count, loff_t *ppos)
{
	return count;
}

static loff_t rust_null_llseek(struct file *file, loff_t offset, int orig)
{
	return file->f_pos = 0;
}

static ssize_t rust_null_read_iter(struct kiocb *iocb, struct iov_iter *to)
{
	return 0;
}

static ssize_t rust_null_write_iter(struct kiocb *iocb, struct iov_iter *from)
{
	size_t count = iov_iter_count(from);

	iov_iter_advance(from, count);
	return count;
}

static const struct file_operations rust_null_fops = {
	.owner		= THIS_MODULE,
	.llseek		= rust_null_llseek,
	.read		= rust_null_read,
	.write		= rust_null_write,
	.read_iter	= rust_null_read_iter,
	.write_iter	= rust_null_write_iter,
};

static struct miscdevice rust_null_misc = {
	.minor = MISC_DYNAMIC_MINOR,
	.name  = "rust-null",
	.fops  = &rust_null_fops,
	.mode  = 0666,
};

static int __init rust_null_init(void)
{
	int err;

	err = misc_register(&rust_null_misc);
	if (err)
		return err;

	pr_info("rust_null: /dev/rust-null registered (misc dynamic minor)\n");
	return 0;
}

static void __exit rust_null_exit(void)
{
	misc_deregister(&rust_null_misc);
	pr_info("rust_null: /dev/rust-null unregistered\n");
}

module_init(rust_null_init);
module_exit(rust_null_exit);

MODULE_LICENSE("GPL");
MODULE_AUTHOR("evie <evie@hareai.dev>");
MODULE_DESCRIPTION("misc device with the /dev/null I/O contract");
MODULE_VERSION("0.1.0");

File: kernel/rfl/rust_null.rs
// SPDX-License-Identifier: GPL-2.0

//! Rust-for-Linux `MiscDevice` with the `/dev/null` contract.
//!
//! This file is the out-of-tree module you would build against a
//! `CONFIG_RUST=y` kernel (`make LLVM=1` in that tree, `M=` this
//! directory, `obj-m := rust_null.o` compiling *this* `.rs`).
//!
//! Distro kernels often ship `CONFIG_HAVE_RUST=y` but not `CONFIG_RUST`.
//! They cannot compile this file. The loadable module in that case is
//! [`rust_null.c`](../rust_null.c), which registers the same
//! `struct miscdevice` name (`rust-null`) and the same
//! `file_operations` (`read` / `write` / `read_iter` / `write_iter` /
//! `llseek`).
//!
//! `kernel::miscdevice::MiscDevice` (mainline) currently hooks
//! `read_iter` / `write_iter` / `ioctl` / `mmap`. It does **not**
//! expose `.llseek`. `/dev/null` uses `null_lseek` (`f_pos = 0`).
//! The C module implements that; a future RfL `llseek` hook should
//! match it. Until then, a pure `MiscDevice` would not pass the seek
//! half of `tests/compare_devices.rs`.

use kernel::{
    fs::{File, Kiocb},
    iov::{IovIterDest, IovIterSource},
    miscdevice::{MiscDevice, MiscDeviceOptions, MiscDeviceRegistration},
    prelude::*,
};

module! {
    type: RustNullModule,
    name: "rust_null",
    authors: ["evie"],
    description: "misc device with the /dev/null I/O contract",
    license: "GPL",
}

#[pin_data]
struct RustNullModule {
    #[pin]
    _miscdev: MiscDeviceRegistration<RustNull>,
}

impl kernel::InPlaceModule for RustNullModule {
    fn init(_module: &'static ThisModule) -> impl PinInit<Self, Error> {
        pr_info!("rust_null: registering misc device rust-null\n");

        let options = MiscDeviceOptions { name: c"rust-null" };

        try_pin_init!(Self {
            _miscdev <- MiscDeviceRegistration::register(options),
        })
    }
}

/// Per-open state. `/dev/null` has none; neither do we.
struct RustNull;

#[vtable]
impl MiscDevice for RustNull {
    type Ptr = ();

    fn open(_file: &File, _misc: &MiscDeviceRegistration<Self>) -> Result<Self::Ptr> {
        Ok(())
    }

    /// `read_null` / `read_iter_null`: always EOF. Do not fill `iov`.
    fn read_iter(_kiocb: Kiocb<'_, Self::Ptr>, _iov: &mut IovIterDest<'_>) -> Result<usize> {
        Ok(0)
    }

    /// `write_null` / `write_iter_null`: accept every byte, discard it.
    fn write_iter(_kiocb: Kiocb<'_, Self::Ptr>, iov: &mut IovIterSource<'_>) -> Result<usize> {
        let count = iov.len();
        iov.advance(count);
        Ok(count)
    }
}
